packages/permission-log-controller/src/PermissionLog.ts

```diff
@@ -0,0 +1,504 @@
+      // Special handling for eth_accounts, in order to record the time the
+      // accounts were last seen or approved by the origin.
+      newEntries = (result as Permission[])
+        .map((perm) => {
+          if (perm.parentCapability === 'eth_accounts') {
+            accounts = this.getAccountsFromPermission(perm);
+          }
+
+          return perm.parentCapability;
+        })
+        .reduce((acc: PermissionEntry, method) => {
+          // all approved permissions will be included in the response,
+          // not just the newly requested ones
+          if (requestedMethods.includes(method)) {
+            if (method === 'eth_accounts') {
+              const accountToTimeMap = getAccountToTimeMap(accounts, time);
+
+              acc[method] = {
+                lastApproved: time,
+                accounts: accountToTimeMap,
+              };
+            } else {
+              acc[method] = { lastApproved: time };
+            }
+          }
+
+          return acc;
+        }, {});
+    }
+
+    if (Object.keys(newEntries).length > 0) {
+      this.commitNewHistory(origin, newEntries);
+    }
+  }
+
+  /**
+   * Commit new entries to the permissions history log.
+   * Merges the history for the given origin, overwriting existing entries
+   * with the same key (permission name).
+   *
+   * @param origin - The requesting origin.
+   * @param newEntries - The new entries to commit.
+   */
+  commitNewHistory(
+    origin: string,
+    newEntries: Record<PermissionName, Partial<PermissionLog>>,
+  ) {
+    // a simple merge updates most permissions
+    const history = this.getHistory();
+    const newOriginHistory = {
+      ...history[origin],
+      ...newEntries,
+    };
+
+    // eth_accounts requires special handling, because of information
+    // we store about the accounts
+    const existingEthAccountsEntry =
+      history[origin] && history[origin].eth_accounts;
+    const newEthAccountsEntry = newEntries.eth_accounts;
+
+    if (existingEthAccountsEntry && newEthAccountsEntry) {
+      // we may intend to update just the accounts, not the permission
+      // itself
+      const lastApproved =
+        newEthAccountsEntry.lastApproved ||
+        existingEthAccountsEntry.lastApproved;
+
+      // merge old and new eth_accounts history entries
+      newOriginHistory.eth_accounts = {
+        lastApproved,
+        accounts: {
+          ...existingEthAccountsEntry.accounts,
+          ...newEthAccountsEntry.accounts,
+        },
+      };
+    }
+
+    this.updateHistory({
+      ...history,
+      [origin]: newOriginHistory as PermissionEntry,
+    });
+  }
+
+  /**
+   * Get all requested methods from a permissions request.
+   *
+   * @param request - The request object.
+   * @returns The names of the requested permissions.
+   */
+  getRequestedMethods(request: JsonRpcRequestWithOrigin<any>): string[] | null {
+    if (
+      !request.params ||
+      !request.params[0] ||
```

Comment:
I see why you're using `any` here. It seems that below, we are assuming that if `request.params` is present, it is an array. However, we can't do that, because it may actually be an object.

To get around this, it seems that we just need an extra check:

``` suggestion
  getRequestedMethods(request: JsonRpcRequestWithOrigin): string[] | null {
    if (
      !request.params ||
      !Array.isArray(request.params) ||
      !request.params[0] ||
```

This does introduce a difference from the original JavaScript code — in fact, we'd be fixing a bug — but I think that's an acceptable change. What do you think?